[Console] Made output docopt compatible

wouterj opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #6329 |
| License | MIT |
| Doc PR | https://github.com/symfony/symfony-docs/issues/5016 |

This was harder than I thought. To sum up:
- The output now follows the [docopt](http://docopt.org/) specification
- There is a new `addUsage` method to add more usage patterns
- The handling of spaces in the descriptors is refactored to make it easier to understand and to make it render better (using sprintf's features only made it worse imo)
## Todo
- [x] Add test for `addUsage` and friends
- [x] Add test for multiline descriptions of arguments
- <s>Convert long descriptions to multiline automatically</s>
- [ ] Submit a doc PR for `addUsage`
## Question

The docopt specification suggests we should add these usage patterns:

```
%command.name% -h | --help
%command.name% --version
```

I didn't do that yet, as I think it'll only makes the output more verbose and it's already pretty obvious.

I've taken some decisions which I don't think everybody agrees with. I'm willing to change it, so feel free to comment :)

/cc @Seldaek

wouterj committed:
Added addUsage() to specify usage patterns

wouterj committed:
Made output docopt compatible

javiereguiluz commented:
@WouterJ could you please post two screenshots of the before/after changes for a typical Symfony command? Thanks!

WouterJ commented:
@javiereguiluz this PR is only about the help output (docopt is only about having a syntax to describe the command). A screenshot when running `server:start --help`: (please note that only the first lines of the help message are shown)

**Before**
![sf-docopt-before](https://cloud.githubusercontent.com/assets/749025/5602820/83663b72-9367-11e4-97ab-4b9d25e78436.png)

**After**
![sf-docopt-after](https://cloud.githubusercontent.com/assets/749025/5602822/8928ade2-9367-11e4-9447-0fe0225203df.png)

javiereguiluz commented:
@WouterJ thanks for the screenshots.

I find the new style way less readable than the original one. I only like the square brackets for the default values.

iltar commented:
I'm not familiar with the standard, but I agree with @javiereguiluz, finding the available options is far easier in the first screenshot. I think this is mainly because when I look at the command, I don't know the available options thus the shortcuts won't make any sense to me until I know the command.

WouterJ commented:
Hmm, let's take a closer look then at what's changed, why it is that way and how we can integrate this in Symfony commands.

First of all, please note that docopt is based on unix commands. It looks very similair, but has some different things. Maybe we would like that more? The differences are mostly in the command synopsis. Now with the `addUsage` method, I start to think that we maybe should use that default as well. An example:

````
$ mv --help
Usage: mv [OPTION]... [-T] SOURCE DEST
  or:  mv [OPTION]... SOURCE... DIRECTORY
  or:  mv [OPTION]... -t DIRECTORY SOURCE...
Rename SOURCE to DEST, or move SOURCE(s) to DIRECTORY.

Mandatory arguments to long options are mandatory for short options too.
      --backup[=CONTROL]       make a backup of each existing destination file
  -b                           like --backup but does not accept an argument
  -f, --force                  do not prompt before overwriting
  -i, --interactive            prompt before overwrite
  -n, --no-clobber             do not overwrite an existing file
````

The Command Synopsis
---

| Before | After | |
| --- | --- | --- |
| `-r|--router="..."` | `-r ROUTER|--router ROUTER` | This is a very common standard (see unix example). However, it'll be more usefull if we can edit the value of the value, having something like `-r PATH|--router=PATH` helps explaining the option without description.<br>Putting the value on both options is maybe more a docopt implementation detail, we can revert that. |
| `argument` | `<argument>` | This makes a clear separation between the command name and arguments |
| `--option argument` | `--option [--] <argument>` | Symfony supports `[--]` and as such, it should show it following the docopt standards. This is also usefull, as not many people know if `[--]` is supported and what it does, this'll help |

We could default this to just:

    Usage: server:start [options] [<address>]

This is valid docopt, as long as the options are defined in the option description and more specific usages can be set using `addUsage()`.

The Options List
---

| Before | After | |
| --- | --- | --- |
| `--router (-r)` | `-r, --router` | This is in the same style as unix and docopt. Consistency is better. Docopt doesn't specify if the shortcut has to be the first, so we can also use `--router, -r`. Moreover, we can tweak it to be indented like Unix commands. |
| `--env (-e)` | `-e=E | --env=E` | For the values, reference to the row in the synopsis about alternatives | `(default: "dev")` | `[default: "dev"]` | Just to comply with docopt standards |

Following this and using some of the things in Unix commands. What do you think of this proposal (not yet implemented):

![sf-docopt-alternative-1](https://cloud.githubusercontent.com/assets/749025/5604134/371b876c-93a9-11e4-8114-019d598ba3aa.png)

javiereguiluz commented:
@WouterJ thanks for the thorough explanation. This new proposal looks definitely much better to me.

iltar commented:
Would it be possible to use a borderless table here? It would align properly, even if there are a lot of options such as with -vvv. The new proposal looks very clear to me as the `--options` are all aligned.

WouterJ commented:
There is no need for a borderless table here. I've not aligned `-v|vv|vvv` with the rest on purpose. As it's such a big option, aligning the complete complete block with it add much whitespace for no good reason. I've moved it to the bottom, so it isn't really annoying imo.

iltar commented:
What happens if an option has a very long description and has a very long default value? An example could be an IPv6 address.

WouterJ commented:
You can add new lines to the description, resulting in:

````
  -l, --long  A very looong description which eventually needs
                 a line break [default: ...........................]

WouterJ commented:
I've now squashed, rebased and improved this PR. It now uses the latest proposed version by me. For the first time, the tests pass (this should however be fixed really soon, it took 4 days to get the tests pass).

I've updated the todo items in the PR description.

Seldaek commented:
Great, thanks @WouterJ :)

wouterj committed:
Customize usage section

fabpot commented:
@WouterJ I can see many todo items in the description, os what's the status of the PR? I had the feeling it was finished, but seeing the todo list, I'm not very sure.

wouterj committed:
Fixed some more tests

WouterJ commented:
@fabpot the feature itself is finished, but the tests are a pain to update (as there is soo much duplication and tests that shouldn't do these assertions). I think the Symfony codebase would benefit from a test cleanup :)

However, this one is at the top of my "finish it" list for this week.

Btw, does anyone know why Travis didn't run for the last commit? My PC has problems executing the suite (windows...), so I rely mostly on Travis for this PR.

fabpot commented:
@WouterJ I agree with you.

It happens to me from time to time as well, Travis do not run anymore for some of my PRs and I don't see how I can force it to run them again. Anyone knows what happen?

iltar commented:
To me it sounds like the commit hook is failing, maybe a connection issue to travis?

javiereguiluz commented:
In case it's useful, this is the official Travis documentation regarding [why PRs aren't built in some cases](http://docs.travis-ci.com/user/pull-requests/#My-Pull-Request-isn't-being-built):

> **My Pull Request isn't being built**
>
> If a pull request isn't built or doesn't show up in Travis CI's user interface, that usually means that it can't be merged. We rely on the merge commit that GitHub transparently creates between the changes in the source branch and the upstream branch the pull request is sent against.
> 
> So when you create or update a pull request, and Travis CI doesn't create a build for it, make sure the pull request is mergeable. If it isn't, rebase it against the upstream branch and resolve any merge conflicts. When you push the fixes up to GitHub and to the pull request, Travis CI will happily test them.
>
> Travis CI also currently doesn't build pull requests when the upstream branch is updated. When this happens, GitHub will update the merge commit between the downstream and upstream branch, and send out a notifications. But Travis CI currently ignores this update, as it could lead to a large number of new builds on repositories with lots of pull requests and lots of updates on the upstream branches.

WouterJ commented:
I'll resolve the merge conflict, maybe that solves the issue. Thanks!

wouterj committed:
Added test for multiline descriptions

wouterj committed:
Made fabbot happy

henrikbjorn commented:
:+1: great idea using docopt and having something that is more unixy

WouterJ commented:
@fabpot this PR is ready for review/merge. The test failures are not caused by this PR or at least, I can't see how this should cause the failures.
I've created a doc issue instead of a PR, as I'm going to revisit the Console docs soon.

WouterJ commented:
ping

WouterJ commented:
ping @symfony/deciders

aitboudad commented:
ping @symfony/deciders

fabpot commented:
Thank you @WouterJ.

rquadling commented on src/Symfony/Component/Console/Descriptor/TextDescriptor.php:
```diff
@@ -52,18 +55,33 @@ protected function describeInputArgument(InputArgument $argument, array $options
+            $synopsis,
+            str_repeat(' ', $spacingWidth),
+            // + 17 = 2 spaces + <info> + </info> + 2 spaces
+            preg_replace('/\s*\R\s*/', "\n".str_repeat(' ', $totalWidth + 17), $option->getDescription()),
```
The additional lines only need to have +4 spaces as the `<info></info>` elements only apply to the first line and that's catered for in the sprintf.

rquadling commented on src/Symfony/Component/Console/Descriptor/TextDescriptor.php:
```diff
@@ -32,16 +32,19 @@ class TextDescriptor extends Descriptor
-            str_replace("\n", "\n".str_repeat(' ', $nameWidth + 2), $argument->getDescription()),
+            str_repeat(' ', $spacingWidth),
+            // + 17 = 2 spaces + <info> + </info> + 2 spaces
+            preg_replace('/\s*\R\s*/', PHP_EOL.str_repeat(' ', $totalWidth + 17), $argument->getDescription()),
```
The additional lines only need to have +4 spaces as the `<info></info>` elements only apply to the first line and that's catered for in the sprintf.
